File: lab2/warshall.py
# floyd warshall algorith in python

def floyd_warshall(graph):
    # number of vertices in the graph
    V = len(graph)
    
    # initialize distance matrix with graph values
    dist = [[float('inf')] * V for _ in range(V)]
    
    for i in range(V):
        for j in range(V):
            dist[i][j] = graph[i][j]
    
    # set distance to self as 0
    # for i in range(V):
    #     dist[i][i] = 0
    
    # update distances using Floyd-Warshall algorithm
    for k in range(V):
        for i in range(V):
            for j in range(V):
                if i == j:
                    dist[i][i] = 0
                else:
                    if dist[i][k] + dist[k][j] < dist[i][j]:
                        dist[i][j] = dist[i][k] + dist[k][j]
    
    return dist

def print_solution(dist):
    V = len(dist)
    for i in range(V):
        for j in range(V):
            if dist[i][j] == float('inf'):
                print("INF", end="\t")
            else:
                print(dist[i][j], end="\t")
        print()

# Example usage
if __name__ == "__main__":
    graph = [
        [0, 3, float('inf'), 7],
        [8, 0, 2, float('inf')],
        [5, float('inf'), 0, 1],
        [2, float('inf'), float('inf'), 0]
    ]
    
    dist = floyd_warshall(graph)
    print("The shortest distance matrix is:")
    print_solution(dist)
    

File: lab2/warshall.c
// floyd warshall algorithm in c to find closure of a graph.
// The graph is represented as an adjacency matrix.
// If there is an edge from i to j, then graph[i][j] is 1, else 0.
// The closure of the graph is represented as a matrix reach[][] such that
// reach[i][j] is 1 if there is a path from i to j, else 0.

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int max(a,b){
    if(a>b)
    return a;
    else
    return b;
}

void warshall(int graph[][10], int n, int reach[][10]) {
    int i, j, k;
    // Initialize the reachability matrix with the given graph
    for (i = 0; i < n; i++)
        for (j = 0; j < n; j++)
            reach[i][j] = graph[i][j];

    // Update the reachability matrix
    for (j = 0; j < n; j++) {
        for (k = 0; k < n; k++) {
            for (i = 0; i < n; i++) {
                reach[i][j] = max(reach[i][j], (reach[i][k] && reach[k][j]));
            }
        }
    }
}

void main(){
    int graph[10][10], reach[10][10];
    int n, i, j;

    printf("Enter the number of vertices in the graph: ");
    scanf("%d", &n);

    printf("Enter the adjacency matrix of the graph (0 or 1):\n");
    for (i = 0; i < n; i++)
        for (j = 0; j < n; j++)
            scanf("%d", &graph[i][j]);

    warshall(graph, n, reach);

    printf("The transitive closure of the graph is:\n");
    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++) {  
            printf("%d ", reach[i][j]);
        }
        printf("\n");
    }
}